src/cpu-template-helper/src/fingerprint/compare.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use clap::ValueEnum;
+use serde::Serialize;
+use vmm::cpu_config::templates::CustomCpuTemplate;
+
+use crate::fingerprint::{Fingerprint, FingerprintField};
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// Difference detected between source and target.
+    #[error("Difference detected between source and target: {0}")]
+    DiffDetected(String),
+    /// Failed to serialize/deserialize JSON.
+    #[error("Failed to serialize/deserialize JSON: {0}")]
+    Serde(#[from] serde_json::Error),
+}
+
+#[derive(Serialize)]
+struct Diff<T: Serialize> {
+    name: String,
+    prev: T,
+    curr: T,
+}
+
+pub fn compare(
+    prev: Fingerprint,
+    curr: Fingerprint,
+    filters: Vec<FingerprintField>,
+) -> Result<(), Error> {
+    let prev = serde_json::to_value(prev)?;
+    let curr = serde_json::to_value(curr)?;
+
+    let mut diff_str = String::new();
+    for filter in filters {
+        let field = filter.to_possible_value().unwrap();
+        let name = field.get_name();
+
+        if prev[name] != curr[name] {
+            if name == "guest_cpu_config" {
+                let cpu_configs: Vec<CustomCpuTemplate> = vec![
+                    serde_json::from_value(prev[name].clone())?,
+                    serde_json::from_value(curr[name].clone())?,
+                ];
+
+                // This `strip()` call always succeed since the number of inputs is two.
+                let cpu_configs = crate::template::strip::strip(cpu_configs).unwrap();
+
+                let diff = Diff {
+                    name: name.to_string(),
+                    prev: cpu_configs[0].clone(),
+                    curr: cpu_configs[1].clone(),
+                };
+                diff_str.push('\n');
+                diff_str.push_str(&serde_json::to_string_pretty(&diff)?);
+            } else {
+                let diff = Diff {
+                    name: name.to_string(),
+                    prev: prev[name].clone(),
+                    curr: curr[name].clone(),
+                };
+                diff_str.push('\n');
+                diff_str.push_str(&serde_json::to_string_pretty(&diff)?);
+            };
+        }
```

Comment:
I am not a fan of this conversion to json (and back). I understand that you did it because of `filters` argument, but I think it will be better to just use `switch` here.
```rust
let compare = |val1, val2, diff| {
    if val1 != val2 { diff.push(Diff {...})};
};
for filter in filters {
    switch filter {
        FirecrackerVersion => {
            compare(prev.version, curr.version);
        }
        ....
    }
}
```

Reply:
So you like this more?
```
pub fn compare(
    prev: Fingerprint,
    curr: Fingerprint,
    filters: Vec<FingerprintField>,
) -> Result<(), Error> {
    let mut diff_str = String::new();
    let mut compare = |field: FingerprintField, val1, val2| -> Result<(), Error> {
        if val1 != val2 {
            let field = field.to_possible_value().unwrap();
            let diff = Diff {
                name: field.get_name().to_string(),
                prev: val1,
                curr: val2,
            };
            diff_str.push('\n');
            diff_str.push_str(&serde_json::to_string_pretty(&diff)?);
        }
        Ok(())
    };

    for filter in filters {
        match filter {
            FingerprintField::firecracker_version => {
                compare(
                    FingerprintField::firecracker_version,
                    prev.firecracker_version.clone(),
                    curr.firecracker_version.clone(),
                );
            }
            FingerprintField::kernel_version => {
                compare(
                    FingerprintField::firecracker_version,
                    prev.kernel_version.clone(),
                    curr.kernel_version.clone(),
                );
            }
            FingerprintField::microcode_version => {
                compare(
                    FingerprintField::firecracker_version,
                    prev.microcode_version.clone(),
                    curr.microcode_version.clone(),
                );
            }
            FingerprintField::bios_version => {
                compare(
                    FingerprintField::firecracker_version,
                    prev.bios_version.clone(),
                    curr.bios_version.clone(),
                );
            }
            FingerprintField::bios_revision => {
                compare(
                    FingerprintField::firecracker_version,
                    prev.bios_revision.clone(),
                    curr.bios_revision.clone(),
                );
            }
            FingerprintField::guest_cpu_config => {
                if prev.guest_cpu_config != curr.guest_cpu_config {
                    let cpu_configs =
                        vec![prev.guest_cpu_config.clone(), curr.guest_cpu_config.clone()];

                    // This `strip()` call always succeed since the number of inputs is two.
                    let cpu_configs = crate::template::strip::strip(cpu_configs).unwrap();

                    let diff = Diff {
                        name: "guest_cpu_config".to_string(),
                        prev: cpu_configs[0].clone(),
                        curr: cpu_configs[1].clone(),
                    };
                    diff_str.push('\n');
                    diff_str.push_str(&serde_json::to_string_pretty(&diff)?);
                }
            }
        }
    }

    if diff_str.is_empty() {
        Ok(())
    } else {
        Err(Error::DiffDetected(diff_str))
    }
}
```

The above still getting the following error, though.
```
error[E0499]: cannot borrow `diff_str` as mutable more than once at a time
  --> src/cpu-template-helper/src/fingerprint/compare.rs:97:21
   |
33 |     let mut compare = |field: FingerprintField, val1, val2| -> Result<(), Error> {
   |                       ---------------------------------------------------------- first mutable borrow occurs here
...
41 |             diff_str.push('\n');
   |             -------- first borrow occurs due to use of `diff_str` in closure
...
50 |                 compare(
   |                 ------- first borrow later used here
...
97 |                     diff_str.push('\n');
   |                     ^^^^^^^^^^^^^^^^^^^ second mutable borrow occurs here

error[E0499]: cannot borrow `diff_str` as mutable more than once at a time
  --> src/cpu-template-helper/src/fingerprint/compare.rs:98:21
   |
33 |     let mut compare = |field: FingerprintField, val1, val2| -> Result<(), Error> {
   |                       ---------------------------------------------------------- first mutable borrow occurs here
...
41 |             diff_str.push('\n');
   |             -------- first borrow occurs due to use of `diff_str` in closure
...
50 |                 compare(
   |                 ------- first borrow later used here
...
98 |                     diff_str.push_str(&serde_json::to_string_pretty(&diff)?);
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ second mutable borrow occurs here

For more information about this error, try `rustc --explain E0499`.
```

Reply:
had an offline discussion and applied the change.